### PR TITLE
Constant propagate float to string

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/regression/jbmc-strings/ConstantEvaluationToString01/Main.java [hunk 1]
```diff
@@ -78,4 +78,19 @@ public void testLongToStringNoPropagation2(int radix) {
     CProver.assume(radix == 2 || radix == 8 || radix == 10 || radix == 16);
     assert CProverString.toString(0L, radix).equals("0");
   }
+
+  public void testFloatToString() {
+    assert CProverString.toString(0F).equals("0.0");
```
Comment (smowton) [inline jbmc/regression/jbmc-strings/ConstantEvaluationToString01/Main.java:83]: Negative zero is the other common edge case, if you can produce that

File: src/goto-symex/goto_symex.cpp [hunk 2]
```diff
@@ -588,3 +595,301 @@ bool goto_symext::constant_propagate_integer_to_string(
...
+    return f.get_sign() ? "-Infinity" : "Infinity";
+  }
+
+  float f_abs = fabsf(f.to_float());
+
+  // remove trailing zeros, except a zero following the decimal point
+  const auto remove_trailing_zeros = [](std::string &s)
+  {
+    std::size_t i;
+    for(i = s.length() - 1; s[i] == '0'; i--);
```
Comment (smowton) [inline src/goto-symex/goto_symex.cpp]: Use `std::string::find_last_not_of`?

File: src/goto-symex/goto_symex.cpp [hunk 3]
```diff
@@ -588,3 +595,301 @@ bool goto_symext::constant_propagate_integer_to_string(
...
+    return f.get_sign() ? "-0.0" : "0.0";
+  }
+  else if(f.is_infinity())
+  {
+    return f.get_sign() ? "-Infinity" : "Infinity";
+  }
+
+  float f_abs = fabsf(f.to_float());
+
+  // remove trailing zeros, except a zero following the decimal point
```
Comment (smowton) [inline src/goto-symex/goto_symex.cpp:1354]: Perhaps note that the input should always contain a decimal point

File: src/goto-symex/goto_symex.cpp [hunk 4]
```diff
@@ -588,3 +595,301 @@ bool goto_symext::constant_propagate_integer_to_string(
 
   return true;
 }
+
+int goto_symext::get_min_precision(const std::string &f1, const std::string &f2)
```
Comment (smowton) [inline src/goto-symex/goto_symex.cpp:1172]: Give some brief commentary about your strategy here

**Review by smowton — COMMENTED** [review]
Of course the min-precision routines look a bit vulnerable to subtle bugs. Can you fuzz this against a real Java implementation to convince us it's right?

File: unit/goto-symex/goto_symex.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    f.from_float(43479347.4f);
+    REQUIRE(goto_symex_testt::float_to_java_string(f) == "4.347935E7");
+
+    f.from_float(-123.456);
+    REQUIRE(goto_symex_testt::float_to_java_string(f) == "-123.456");
+
+    f.from_float(-43479347.4f);
+    REQUIRE(goto_symex_testt::float_to_java_string(f) == "-4.347935E7");
+  }
```
Comment (martin-cs) [inline unit/goto-symex/goto_symex.cpp:77]: I would also suggest, `-NaN`, `1/3`, `1/7`, `1/10` and `2.2250738585072012e-308` (see https://www.exploringbinary.com/java-hangs-when-converting-2-2250738585072012e-308/ )

**Review by martin-cs — COMMENTED** [review]
What is your criteria for correctness of this code?  There are several algorithms for doing this conversion, each of which have their own ... quirks (see the link in the comment) so if your criteria for correctness is "must match the behaviour of ____ platform exactly" then you may need more evidence that this is right.

**Comment by TGWDB** [comment]
Closing as this is orphaned and no sign of progress in years. Please reopen if you believe this is erroneous and wish to bring this up to date.

**Pull request closed without merging**